autoprecompiles/src/memory_optimizer.rs

```diff
@@ -0,0 +1,323 @@
+use powdr_constraint_solver::range_constraint::RangeConstraint;
+use powdr_constraint_solver::{
+    quadratic_symbolic_expression::QuadraticSymbolicExpression,
+    symbolic_expression::SymbolicExpression,
+};
+use powdr_number::{FieldElement, LargeInt};
+
+use crate::{MemoryBusInteraction, MemoryType, SymbolicConstraint, SymbolicMachine};
+
+pub fn optimize_memory<T: FieldElement>(mut machine: SymbolicMachine<T>) -> SymbolicMachine<T> {
+    let memory_bus_interactions = machine
+        .bus_interactions
+        .iter()
+        .filter_map(|bus| {
+            let mem_int = MemoryBusInteraction::try_from(bus.clone()).ok()?;
+            if matches!(mem_int.ty, MemoryType::Memory) {
+                Some(mem_int)
+            } else {
+                None
+            }
+        })
+        .collect_vec();
+
+    // Transform the constraints such that some quadratic constraints are turned into
+    // affine when we introduce a new boolean variable ("zero check variable").
+    let mut zero_check_transformer = ZeroCheckTransformer::default();
+    let constraints = machine
+        .constraints
+        .iter()
+        .map(|constr| {
+            let constr = algebraic_to_quadratic_symbolic_expression(&constr.expr);
+            let constr = try_remove_is_valid(&constr).unwrap_or(&constr);
+            zero_check_transformer.transform(constr.clone())
+        })
+        .collect_vec();
+
+    let constraints_by_variable = constraints
+        .iter()
+        .flat_map(|constr| {
+            constr
+                .referenced_unknown_variables()
+                .map(move |var| (var.clone(), constr))
+        })
+        .into_group_map();
+
+    // Collect the expressions that are used as addresses and try to solve
+    // a constraint about them in the constraint system.
+    let memory_addresses = memory_bus_interactions
+        .iter()
+        .map(|bus| algebraic_to_quadratic_symbolic_expression(&bus.addr))
+        .flat_map(|addr| {
+            // Go through the constraints related to this address
+            // and try to solve for the address.
+            let expr = addr
+                .referenced_unknown_variables()
+                .flat_map(|v| constraints_by_variable.get(v).map(|constrs| constrs.iter()))
+                .flatten()
+                .unique_by(|constr| constr as *const _)
+                .find_map(|constr| constr.try_solve_for_expr(&addr));
+            // If we can solve for the address, we just take the address unmodified.
+            Some((addr.clone(), expr.unwrap_or(addr)))
+        })
+        .collect::<BTreeMap<_, _>>();
+
+    let mut new_constraints: Vec<SymbolicConstraint<T>> = Vec::new();
+
+    // Go through the memory interactions and try to optimize them.
+    let mut memory_contents: BTreeMap<_, (Option<usize>, Vec<AlgebraicExpression<_>>)> =
+        BTreeMap::new();
+    let mut to_remove: BTreeSet<usize> = Default::default();
+    let mut is_receive = false;
```

Comment:
This could be `zip([false, true].iter().cycle())` to avoid the mutable variable.
Also, what guarantees that memory interactions are alternating sends and receives?

Reply:
I copied it from the register optimizer, so probably @leonardoalt can answer better.